x-pack/plugin/analytics/src/main/java/org/elasticsearch/xpack/analytics/multiterms/MultiTermsAggregator.java

```diff
@@ -0,0 +1,538 @@
+        if (subAggsNeedScore() && descendsFromNestedAggregator(parent)) {
+            /**
+             * Force the execution to depth_first because we need to access the score of
+             * nested documents in a sub-aggregation and we are not able to generate this score
+             * while replaying deferred documents.
+             */
+            this.collectMode = SubAggCollectionMode.DEPTH_FIRST;
+        } else {
+            this.collectMode = collectMode;
+        }
+        // TODO: Make a function
+        // Don't defer any child agg if we are dependent on it for pruning results
+        if (order instanceof InternalOrder.Aggregation) {
+            aggsUsedForSorting.add(((InternalOrder.Aggregation) order).path().resolveTopmostAggregator(this));
+        } else if (order instanceof InternalOrder.CompoundOrder) {
+            InternalOrder.CompoundOrder compoundOrder = (InternalOrder.CompoundOrder) order;
+            for (BucketOrder orderElement : compoundOrder.orderElements()) {
+                if (orderElement instanceof InternalOrder.Aggregation) {
+                    aggsUsedForSorting.add(((InternalOrder.Aggregation) orderElement).path().resolveTopmostAggregator(this));
+                }
+            }
+        }
+        this.needsScore = configs.stream().anyMatch(c -> c.getValuesSource().needsScores());
+        values = configs.stream()
+            .map(c -> context.getValuesSourceRegistry().getAggregator(REGISTRY_KEY, c).build(c))
+            .collect(Collectors.toList());
+        keyConverters = values.stream().map(TermValuesSource::keyConverter).collect(Collectors.toList());
+        bucketOrds = BytesKeyedBucketOrds.build(context.bigArrays(), cardinality);
+
+    }
+
+    private boolean subAggsNeedScore() {
+        for (Aggregator subAgg : subAggregators) {
+            if (subAgg.scoreMode().needsScores()) {
+                return true;
+            }
+        }
+        return false;
+    }
+
+    @Override
+    public ScoreMode scoreMode() {
+        if (needsScore) {
+            return ScoreMode.COMPLETE;
+        }
+        return super.scoreMode();
+    }
+
+    @Override
+    protected boolean shouldDefer(Aggregator aggregator) {
+        return collectMode == SubAggCollectionMode.BREADTH_FIRST && !aggsUsedForSorting.contains(aggregator);
+    }
+
+    List<TermValues> termValuesList(LeafReaderContext ctx) throws IOException {
+        List<TermValues> termValuesList = new ArrayList<>();
+        for (TermValuesSource termValuesSource : values) {
+            termValuesList.add(termValuesSource.getValues(ctx));
+        }
+        return termValuesList;
+    }
+
+    List<List<Object>> docTerms(List<TermValues> termValuesList, int doc) throws IOException {
+        List<List<Object>> terms = new ArrayList<>();
+        for (TermValues termValues : termValuesList) {
+            List<Object> values = termValues.collectValues(doc);
+            if (values == null) {
+                return null;
```

Comment:
Am I interpreting this correct that if one of the doc fields is missing/null, we basically skip the entire document?

If yes, presumably setting `missing` is the required action if you want the document to be aggregated?  Do we make a note of that anywhere in the docs?

Reply:
In the documentation of missing parameter we have:
> By default if any of the key components are missing the entire document will be ignored
but it is also possible to treat them as if they had a value.

I think it might have been not very clear. So I rephrased it a bit.